### PR TITLE
Analytics for related genes ideogram (SCP-2720)

### PR DESCRIPTION
**Pull request by eweitz**
This instruments Mixpanel logging for Ideogram for related genes (#735).  It affects two events:

* Updated `search` event for gene search in Study Overview.  Searches triggered via ideogram are tracked via the `trigger` property.  In these cases, context about the related genes ideogram is logged in namespaced supplemental properties.

* New `plot:related-genes-ideogram` event, like `plot:violin`.  This tracks the searched gene, number of paralogs and interacting genes returned, and time taken for related genes API calls done internally by Ideogram.

These analytics are designed to help answer questions about this new feature, like:
* Do people use it?
* How many and what kind of related genes are being returned?
* Is it fast?  What affects performance?
* What proportion of study gene searches are done via related genes ideogram?
* What search patterns emerge?  Do people search in a hub-and-spoke manner, or an acyclic path, etc.?

This satisfies SCP-2720.

### CODE REVIEW

File: app/assets/javascripts/scp-related-genes-ideogram.js [hunk 1]
```diff
@@ -13,8 +13,21 @@
 
 /** Handle clicks on Ideogram annotations */
 function onClickAnnot(annot) {
+  const ideo = this // eslint-disable-line
```
Comment (devonbush) [inline app/assets/javascripts/scp-related-genes-ideogram.js:16]: what is the 'this' here?  is it the ideoConfig object?

File: app/assets/javascripts/scp-related-genes-ideogram.js [hunk 2]
```diff
@@ -13,8 +13,21 @@
...
+  // This helps profile the numerator of click-through-rate
+  const namespacedProps = {}
+  const props = getRelatedGenesAnalytics(ideo)
+  Object.entries(props).forEach(([key, value]) => {
+    namespacedProps[`relatedGenes:${key}`] = value
+  })
+
+  const event = Object.assign(
+    { type: 'click-related-genes' }, namespacedProps
+  )
```
Comment (devonbush) [inline app/assets/javascripts/scp-related-genes-ideogram.js:29]:
```suggestion
 namespacedProps['type'] = 'click-related-genes'
```
namespacedProps is a fresh object and you're not using it anywhere else, so no need for the extra copy

File: app/assets/javascripts/scp-related-genes-ideogram.js [hunk 3]
```diff
@@ -13,8 +13,21 @@
...
+  window.submitGeneSearch(event)
```
Comment (devonbush) [inline app/assets/javascripts/scp-related-genes-ideogram.js:30]:
```suggestion
  window.submitGeneSearch(namespacedProps)
```

File: app/assets/javascripts/scp-related-genes-ideogram.js [hunk 4]
```diff
@@ -72,6 +85,31 @@ function showRelatedGenesIdeogram() { // eslint-disable-line
...
+ */
+function getRelatedGenesAnalytics(ideo) {
+  const props = Object.assign({}, ideo.relatedGenesAnalytics)
+
+  // Use DSP-conventional name
+  props['perfTime'] = props.time
+  delete props['time']
+
+  props['species'] = window.SCP.taxon
+  return Object.assign({}, props)
```
Comment (devonbush) [inline app/assets/javascripts/scp-related-genes-ideogram.js:99]: why do we need another copy?  props is a fresh object

File: app/assets/javascripts/scp-related-genes-ideogram.js [hunk 5]
```diff
@@ -72,6 +85,31 @@ function showRelatedGenesIdeogram() { // eslint-disable-line
...
+}
+
+/**
+ * Callback to report analytics to Mixpanel.
+ * Helps profile denominator of click-through-rate
+ */
+function onPlotRelatedGenes() {
+  const ideo = this // eslint-disable-line
```
Comment (devonbush) [inline app/assets/javascripts/scp-related-genes-ideogram.js:107]: seems like you should either use a more descriptive name than 'ideo' (and add a comment about what it is) or just drop this line and use 'this' on the line below.

**Review by devonbush — CHANGES_REQUESTED** [review]

**Comment by devonbush** [comment]
Overall makes sense, just some style quibbles

File: app/assets/javascripts/scp-related-genes-ideogram.js (hunk 4 above)
Comment (eweitz) [reply app/assets/javascripts/scp-related-genes-ideogram.js]: Fixed in dbd7beb.

File: app/assets/javascripts/scp-related-genes-ideogram.js (hunk 1 above)
Comment (eweitz) [reply app/assets/javascripts/scp-related-genes-ideogram.js]: It's the Ideogram object (the config object is accessible via `ideogram.config`, FWIW).  I renamed the variable for clarity and added a brief comment in dbd7beb.

**Review by devonbush — APPROVED** [review]
looks good, although if you can think of a better name than plot-async I'm all ears.

**Review by knapii-developments — APPROVED** [review]
Thank you for walking me through this!

**Comment by eweitz** [comment]
The [build failures](https://travis-ci.com/github/broadinstitute/single_cell_portal_core/builds/190095800) seem unrelated:

```
  1) Failure:
SearchControllerTest#test_should_run_inferred_search_using_facets [test/api/search_controller_test.rb:346]:
Did not find expected accessions before inferred search, expected ["SCP3"] but found [].
Expected: ["SCP3"]
  Actual: []
  2) Failure:
SearchControllerTest#test_should_find_intersection_of_facets_on_inferred_search [test/api/search_controller_test.rb:413]:
Did not find expected accessions before inferred search, expected ["SCP3"] but found [].
Expected: ["SCP3"]
  Actual: []
  3) Failure:
SearchControllerTest#test_should_return_search_results_using_facets [test/api/search_controller_test.rb:118]:
Did not return correct array of matching accessions, expected ["SCP3"] but found [].
Expected: ["SCP3"]
  Actual: []
  4) Failure:
SearchControllerTest#test_should_run_inferred_search_using_facets_and_phrase [test/api/search_controller_test.rb:381]:
Did not find expected accessions for phrase & facet search, expected ["SCP3", "SCP2"] but found ["SCP2"].
Expected: ["SCP3", "SCP2"]
  Actual: ["SCP2"]
  5) Error:
SearchControllerTest#test_should_return_search_results_using_numeric_facets:
NoMethodError: undefined method `+' for nil:NilClass
    test/api/search_controller_test.rb:141:in `block (2 levels) in <class:SearchControllerTest>'
    test/api/search_controller_test.rb:140:in `each'
    test/api/search_controller_test.rb:140:in `block in <class:SearchControllerTest>'
  6) Error:
SearchControllerTest#test_should_return_preview_of_bulk_download_files_and_total_bytes:
NoMethodError: undefined method `upload_file_size' for nil:NilClass
    test/api/search_controller_test.rb:277:in `block in <class:SearchControllerTest>'
19 runs, 71 assertions, 4 failures, 2 errors, 0 skips
```

**Pull request merged**